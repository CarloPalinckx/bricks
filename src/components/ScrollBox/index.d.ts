import { Component } from 'react';
import { StyledType } from '../../utility/styled';
declare type PropsType = StyledType & {
    autoHideScrollBar?: boolean;
    showInsetShadow?: boolean;
    onScroll?(eventData: {
        scrollTop: number;
        scrollBottom: number;
        scrollDirection: StateType['scrollDirection'];
    }): void;
};
declare type StateType = {
    scrollPosition: number;
    scrollDirection: 'up' | 'down' | 'idle';
    showInsetShadow: boolean;
};
declare class ScrollBox extends Component<PropsType, StateType> {
    private scrollbar;
    private contentRef;
    constructor(props: PropsType);
    private handleScroll;
    private hasOverflow;
    componentDidMount(): void;
    componentWillUnmount(): void;
    render(): JSX.Element;
}
export default ScrollBox;
export { StateType, PropsType };
