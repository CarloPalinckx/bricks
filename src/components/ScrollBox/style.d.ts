import { default as _R } from 'react';
import { StyledComponentClass as _S } from 'styled-components';
import _T from '../../types/ThemeType';
declare type ScrollBoxThemeType = {
    scrollbar: {
        background: string;
    };
};
declare type effectPropsType = {
    show: boolean;
};
declare const StyledWrapper: _S<_R.DetailedHTMLProps<_R.HTMLAttributes<HTMLDivElement>, HTMLDivElement>, _T, _R.DetailedHTMLProps<_R.HTMLAttributes<HTMLDivElement>, HTMLDivElement>>;
declare const StyledScrollBox: _S<_R.DetailedHTMLProps<_R.HTMLAttributes<HTMLDivElement>, HTMLDivElement>, _T, _R.DetailedHTMLProps<_R.HTMLAttributes<HTMLDivElement>, HTMLDivElement>>;
declare const StyledTop: _S<effectPropsType & _R.HTMLProps<HTMLDivElement>, _T, effectPropsType & _R.HTMLProps<HTMLDivElement>>;
declare const StyledBottom: _S<effectPropsType & _R.HTMLProps<HTMLDivElement>, _T, effectPropsType & _R.HTMLProps<HTMLDivElement>>;
export default StyledScrollBox;
export { StyledWrapper, ScrollBoxThemeType, StyledBottom, StyledTop };
