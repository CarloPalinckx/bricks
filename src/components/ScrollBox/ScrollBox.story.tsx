import { boolean, number } from '@storybook/addon-knobs/react';
import { storiesOf } from '@storybook/react';
import React, { Component } from 'react';
import ScrollBox from '.';
import styled from '../../utility/styled';
import trbl from '../../utility/trbl';
import Heading from '../Heading';
import Spacer from '../Spacer';
import SubHeading from '../SubHeading';
import Text from '../Text';

const demoContent = `
    Lorem ipsum dolor sit amet, consectetur adipiscing elit. Etiam ut mi ligula. Phasellus tellus nulla,
    cursus sit amet dolor eu, sodales facilisis tortor. Maecenas sed arcu quis est pharetra molestie sed
    eu leo. Mauris consequat mauris et eros gravida vestibulum. Phasellus convallis ipsum quis nisl lacinia,
    a pulvinar est porta. Nunc tempus vulputate dapibus. In eget venenatis orci. Pellentesque habitant morbi
    tristique senectus et netus et malesuada fames ac turpis egestas. Maecenas finibus lorem et quam imperdiet,
    id vestibulum odio molestie. Curabitur euismod sit amet tortor et imperdiet. Nam a nisl quis lorem porta
    pharetra. Duis sed magna vel odio ullamcorper gravida eu et nibh. Mauris tempor libero purus, vitae lacinia
    felis hendrerit in. Vestibulum rhoncus vitae ipsum vel placerat.

    Fusce euismod, sem quis tristique dictum, dolor nulla fermentum turpis, sed ultricies augue metus eget
    lorem. Donec nisi leo, tristique ac pretium id, aliquam eget purus. Aenean nec lectus vitae eros condimentum
    maximus et nec metus. Vestibulum ultricies dui non consequat porta. Vestibulum nec scelerisque sapien.
    Nam eu urna nec augue pellentesque tempus. Nulla facilisi. Phasellus fringilla nisi ut lorem aliquet
    mattis. Duis egestas dui ligula, ut lobortis purus suscipit id. Nunc non faucibus arcu. Integer eu
    elit tellus. Fusce suscipit sem sem, ut aliquam elit mollis vel. Sed quis accumsan dolor. Nullam
    gravida nec purus nec facilisis. Lorem ipsum dolor sit amet, consectetur adipiscing elit. Aliquam
    lectus purus, vulputate ut egestas condimentum, condimentum vel lorem.
    Sed vitae consequat eros. Proin in magna odio. Morbi tempus nulla leo, sed condimentum velit semper et.
    Duis porttitor odio nibh, eget pellentesque ligula dignissim quis. Vivamus elit lacus, maximus ac rhoncus
    ultricies, dapibus at mi. Fusce sagittis nisl non lorem auctor viverra. Morbi lacinia vestibulum purus,
    at malesuada urna hendrerit sed.

    Proin ultricies vel nisl ac sollicitudin. Cras nisi mi, consectetur et sapien eget,
    placerat pellentesque ex. In sollicitudin tellus id tincidunt facilisis.
    Nunc hendrerit at eros vel venenatis. Pellentesque finibus sem tincidunt auctor cursus.
    Phasellus commodo ornare felis, id suscipit turpis aliquam ac. Donec et metus condimentum,
    dignissim sem et, dictum sapien.

    Cras elementum enim lorem, et pharetra nisi accumsan eu. Praesent
    feugiat nibh lacus, nec viverra turpis mattis et. Sed dolor urna,
    efficitur vitae ultrices eu, faucibus ac libero. Quisque euismod
    elit mattis, tincidunt lorem sed, fermentum risus. Morbi elementum cursus venenatis.
    Ut turpis sem, varius in ultrices quis, tristique quis justo. In tortor neque, mollis
    ut lacinia quis, euismod vel libero. Sed nec ipsum ante. Vivamus facilisis nulla
    facilisis magna porta, tempor congue arcu sodales. Praesent commodo at lacus nec feugiat.
    Donec libero dui, accumsan eget metus non, pharetra euismod erat. Quisque tincidunt est
    in odio laoreet consequat consectetur at lectus. Praesent ut porta leo.
`;

const StyledDiv = styled.div`
    max-width: 300px;
    border: solid 3px rgba(255, 36, 94, 0.3);
`;

type DemoStateType = {
    scrollTop:number;
    scrollBottom?:number;
    scrollDirection:string;
};

class Demo extends Component<{}, DemoStateType> {
    public constructor(props:{}) {
        super(props);

        this.state = {
            scrollTop: 0,
            scrollDirection: 'idle',
        };
    }

    public render():JSX.Element {
        return (
            <>
                <StyledDiv>
                    <ScrollBox
                        maxHeight={number('maxHeight', 400)}
                        showInsetShadow={boolean('showInsetShadow', true)}
                        autoHideScrollBar={boolean('autoHideScrollBar', true)}
                        onScroll={({ scrollTop, scrollBottom, scrollDirection }):void => {
                            this.setState({
                                scrollTop,
                                scrollBottom,
                                scrollDirection,
                            });
                        }}
                    >
                        <Spacer offsetType="inner" offset={trbl(12, 24, 12, 12)}>
                            <Heading>Lorem ipsum dolor sit amet.</Heading>
                            <Text>
                                {demoContent}
                            </Text>
                        </Spacer>
                    </ScrollBox>
                </StyledDiv>
                <Spacer offset={trbl(48, 0, 12)}>
                    <SubHeading>
                        Scroll event
                    </SubHeading>
                </Spacer>
                <Text>
                    scrollTop: {this.state.scrollTop}px<br />
                    scrollBottom: {this.state.scrollBottom}px<br />
                    scrollDirection: {this.state.scrollDirection}
                </Text>
            </>
        );
    }
}

storiesOf('ScrollBox', module)
    .add('Default', () => {
        return (
            <Demo />
        );
    });
