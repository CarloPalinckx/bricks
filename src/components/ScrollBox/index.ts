export { default as default, ScrollBoxThemeType } from './ScrollBox.style';
