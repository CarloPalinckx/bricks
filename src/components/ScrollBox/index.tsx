import React, { Component } from 'react';
import ScrollBar from 'simplebar';
import { StyledType } from '../../utility/_styled';
import StyledScrollBox, { StyledBottom, StyledTop, StyledWrapper } from './style';

type PropsType = StyledType & {
    autoHideScrollBar?: boolean;
    showInsetShadow?: boolean;
    onScroll?(eventData: {
        scrollTop: number;
        scrollBottom: number;
        scrollDirection: StateType['scrollDirection'];
    }): void;
};

type StateType = {
    scrollPosition: number;
    scrollDirection: 'up' | 'down' | 'idle';
    showInsetShadow: boolean;
};

class ScrollBox extends Component<PropsType, StateType> {
    private scrollbar: ScrollBar;
    private contentRef: HTMLDivElement | null;

    public constructor(props: PropsType) {
        super(props);

        this.state = {
            scrollPosition: 0,
            scrollDirection: 'idle',
            showInsetShadow: false,
        };
    }

    private handleScroll = (): void => {
        const element = this.scrollbar.getScrollElement();
        const contentElement = this.scrollbar.getContentElement();
        const direction = this.state.scrollPosition >= element.scrollTop ? 'up' : 'down';

        const scrollBottom =
            (contentElement as HTMLDivElement).offsetHeight -
            element.scrollTop -
            (element as HTMLDivElement).offsetHeight;

        if (this.props.onScroll !== undefined) {
            this.props.onScroll({
                scrollTop: element.scrollTop,
                scrollBottom,
                scrollDirection: direction,
            });
        }

        this.setState({
            showInsetShadow: this.hasOverflow(),
            scrollPosition: element.scrollTop,
            scrollDirection: direction,
        });
    };

    private hasOverflow = (): boolean => {
        const element = this.scrollbar.getScrollElement();
        const contentElement = this.scrollbar.getContentElement();

        return (element as HTMLDivElement).offsetHeight < (contentElement as HTMLDivElement).offsetHeight;
    };

    public componentDidMount(): void {
        this.scrollbar = new ScrollBar(this.contentRef as HTMLDivElement, {
            autoHide: this.props.autoHideScrollBar !== undefined ? this.props.autoHideScrollBar : true,
        });

        this.scrollbar.getScrollElement().addEventListener('scroll', this.handleScroll);

        this.setState({
            showInsetShadow: this.hasOverflow(),
        });
    }

    public componentWillUnmount(): void {
        this.scrollbar.getScrollElement().removeEventListener('scroll', this.handleScroll);
    }

    public render(): JSX.Element {
        return (
            <StyledWrapper>
                <StyledScrollBox
                    innerRef={(ref): void => {
                        this.contentRef = ref;
                    }}
                >
                    {this.props.showInsetShadow !== false && (
                        <StyledTop show={this.state.showInsetShadow && this.state.scrollDirection === 'down'} />
                    )}
                    {this.props.children}
                    {this.props.showInsetShadow !== false && (
                        <StyledBottom show={this.state.showInsetShadow && this.state.scrollDirection !== 'down'} />
                    )}
                </StyledScrollBox>
            </StyledWrapper>
        );
    }
}

export default ScrollBox;
export { StateType, PropsType };
