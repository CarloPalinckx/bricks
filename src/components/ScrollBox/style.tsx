import { default as _R } from 'react';
import { StyledComponentClass as _S } from 'styled-components';
import { PropsType } from '.';
import _T from '../../types/ThemeType';
import styled from '../../utility/styled';

type ScrollBoxThemeType = {
    scrollbar: {
        background: string;
    };
};

const StyledTop = styled.div`
    background: linear-gradient(to bottom, rgba(0, 0, 0, 0.1) 20%, rgba(0, 0, 0, 0) 100%);
    height: 6px;
    position: absolute;
    top: 0;
    left: 0;
    right: 0;
`;

const StyledBottom = styled.div`
    background: linear-gradient(to bottom, rgba(0, 0, 0, 0) 20%, rgba(0, 0, 0, 0.1) 100%);
    height: 6px;
    position: absolute;
    bottom: 0;
    left: 0;
    right: 0;
`;

const StyledScrollBox = styled.div`
    position: relative;
    max-height: ${({ maxHeight }: PropsType): string => `${maxHeight}px`};
    overflow: hidden;

    .simplebar-scrollbar::before {
        background: ${({ theme }): string => theme.ScrollBox.scrollbar.background};
    }
`;

export default StyledScrollBox;
export { StyledTop, StyledBottom, ScrollBoxThemeType };
