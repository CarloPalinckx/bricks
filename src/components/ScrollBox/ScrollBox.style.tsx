import { default as _R } from 'react';
import { StyledComponentClass as _S } from 'styled-components';
import _T from '../../types/ThemeType';
import styled from '../../utility/styled';
import ScrollBox, { PropsType } from './ScrollBox.template';

type ScrollBoxThemeType = {
    scrollbar: {
        background: string;
    };
};

const StyledTop = styled.div`
    background: linear-gradient(
        to bottom,
        rgba(0, 0, 0, 0.1) 20%,
        rgba(0, 0, 0, 0) 100%
    );
    height: 6px;
    position: absolute;
    top: 0;
    left: 0;
    right: 0;
`;

const StyledBottom = styled.div`
    background: linear-gradient(
        to bottom,
        rgba(0, 0, 0, 0) 20%,
        rgba(0, 0, 0, 0.1) 100%
    );
    height: 6px;
    position: absolute;
    bottom: 0;
    left: 0;
    right: 0;
`;

const StyledScrollBox = styled(ScrollBox)`
    position: relative;
    max-height: ${({ maxHeight }: PropsType): string => `${maxHeight}px`};
    overflow: hidden;

    .simplebar-scrollbar::before {
        background: ${({ theme }): string =>
            theme.ScrollBox.scrollbar.background};
    }
`;

export default StyledScrollBox;
export { StyledTop, StyledBottom, ScrollBoxThemeType };
