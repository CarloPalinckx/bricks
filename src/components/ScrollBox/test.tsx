import { mount } from 'enzyme';
import toJson from 'enzyme-to-json';
import React from 'react';
import ScrollBar from 'simplebar';
import ScrollBox from '.';
import MosTheme from '../../themes/MosTheme';

let scrollBarOptions = {};

jest.mock('simplebar', () =>
    jest.fn().mockImplementation((element: HTMLElement, options: Object) => {
        scrollBarOptions = options;

        return {
            getScrollElement: jest.fn().mockReturnValue({
                offsetHeight: 200,
                scrollTop: 0,
                addEventListener: jest.fn().mockImplementation((event, callback) => {
                    callback();
                }),
            }),
            getContentElement: jest.fn().mockReturnValue({
                offsetHeight: 400,
            }),
        };
    }),
);

beforeEach(() => {
    scrollBarOptions = {};
});

describe('ScrollBox', () => {
    it('should show a top shadow when scrolling down', () => {
        /* tslint:disable */
        (ScrollBar as any).mockImplementationOnce(
            (element: HTMLElement, options: Object): Object => {
                /* tslint:enable */
                scrollBarOptions = options;

                return {
                    getScrollElement: jest.fn().mockReturnValue({
                        addEventListener: jest.fn().mockImplementation((event, callback) => {
                            callback();
                        }),
                    }),
                    getContentElement: jest.fn().mockReturnValue({
                        offsetHeight: 400,
                    }),
                };
            },
        );

        const component = mount(
            <MosTheme>
                <ScrollBox showInsetShadow maxHeight={900} />
            </MosTheme>,
        );

        const shadow = component
            .childAt(0)
            .childAt(0)
            .childAt(0)
            .childAt(0)
            .childAt(0);

        /* tslint:disable */
        expect(toJson(shadow)).toMatchSnapshot();
        (expect(shadow) as any).toHaveStyleRule('position', 'absolute');
        (expect(shadow) as any).toHaveStyleRule('top', '0');
        /* tslint:enable */
    });

    it('should show a bottom shadow when scrolling up', () => {
        const component = mount(
            <MosTheme>
                <ScrollBox showInsetShadow maxHeight={900} />
            </MosTheme>,
        );

        const shadow = component
            .childAt(0)
            .childAt(0)
            .childAt(0)
            .childAt(0)
            .childAt(1);

        /* tslint:disable */
        expect(toJson(shadow)).toMatchSnapshot();
        (expect(shadow) as any).toHaveStyleRule('position', 'absolute');
        (expect(shadow) as any).toHaveStyleRule('bottom', '0');
        /* tslint:enable */
    });

    it('should hide the scrollbar when autoHideScrollBar is set', () => {
        mount(
            <MosTheme>
                <ScrollBox autoHideScrollBar maxHeight={900} />
            </MosTheme>,
        );

        expect(scrollBarOptions).toEqual({ autoHide: true });
    });

    it('should handle a scroll event', () => {
        const onScrollMock = jest.fn();

        mount(
            <MosTheme>
                <ScrollBox autoHideScrollBar maxHeight={900} onScroll={onScrollMock} />
            </MosTheme>,
        );

        expect(onScrollMock).toHaveBeenCalledWith({
            scrollBottom: 200,
            scrollDirection: 'up',
            scrollTop: 0,
        });
    });

    it('should tear down any listeners on unmount', () => {
        const removeListenerMock = jest.fn();

        /* tslint:disable */
        (ScrollBar as any).mockImplementationOnce(
            (element: HTMLElement, options: Object): Object => {
                /* tslint:enable */
                return {
                    getScrollElement: jest.fn().mockReturnValue({
                        addEventListener: jest.fn().mockImplementation((event, callback) => {
                            callback();
                        }),
                        removeEventListener: removeListenerMock,
                    }),
                    getContentElement: jest.fn().mockReturnValue({
                        offsetHeight: 400,
                    }),
                };
            },
        );

        const component = mount(
            <MosTheme>
                <ScrollBox autoHideScrollBar maxHeight={900} />
            </MosTheme>,
        );

        component.unmount();

        expect(removeListenerMock).toHaveBeenCalled();
    });
});
