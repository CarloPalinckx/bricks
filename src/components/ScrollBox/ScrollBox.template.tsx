import React, { Component, createRef, RefObject } from 'react';
import ScrollBar from 'simplebar';
import 'simplebar/dist/simplebar.css';
import { StyledType } from '../../utility/styled';
import { StyledBottom, StyledTop } from './ScrollBox.style';

type PropsType = StyledType & {
    maxHeight: number;
    showInsetShadow?: boolean;
    autoHideScrollBar?: boolean;
    onScroll?(eventData: {
        scrollTop: number;
        scrollBottom: number;
        scrollDirection: StateType['scrollDirection'];
    }): void;
};

type StateType = {
    scrollPosition: number;
    scrollDirection: 'up' | 'down';
};

class ScrollBox extends Component<PropsType, StateType> {
    private scrollbar: ScrollBar;
    private contentRef: RefObject<HTMLDivElement>;

    public constructor(props: PropsType) {
        super(props);
        this.contentRef = createRef();

        this.state = {
            scrollPosition: 0,
            scrollDirection: 'up',
        };
    }

    private handleScroll = (): void => {
        const element = this.scrollbar.getScrollElement();

        const contentElement = this.scrollbar.getContentElement();

        const direction =
            this.state.scrollPosition >= element.scrollTop ? 'up' : 'down';

        const scrollBottom =
            (contentElement as HTMLDivElement).offsetHeight -
            element.scrollTop -
            (element as HTMLDivElement).offsetHeight;

        if (this.props.onScroll !== undefined) {
            this.props.onScroll({
                scrollTop: element.scrollTop,
                scrollBottom,
                scrollDirection: direction,
            });
        }

        this.setState({
            scrollPosition: element.scrollTop,
            scrollDirection: direction,
        });
    };

    public componentDidMount(): void {
        this.scrollbar = new ScrollBar(
            this.contentRef.current as HTMLDivElement,
            {
                autoHide:
                    this.props.autoHideScrollBar !== undefined
                        ? this.props.autoHideScrollBar
                        : true,
            },
        );

        this.scrollbar
            .getScrollElement()
            .addEventListener('scroll', this.handleScroll);
    }

    public componentWillUnmount(): void {
        this.scrollbar
            .getScrollElement()
            .removeEventListener('scroll', this.handleScroll);
    }

    public render(): JSX.Element {
        return (
            <div className={this.props.className}>
                {this.props.showInsetShadow === true &&
                this.state.scrollDirection === 'down' ? (
                    <StyledTop />
                ) : (
                    undefined
                )}
                <div ref={this.contentRef}>{this.props.children}</div>
                {this.props.showInsetShadow === true &&
                this.state.scrollDirection === 'up' ? (
                    <StyledBottom />
                ) : (
                    undefined
                )}
            </div>
        );
    }
}

export default ScrollBox;
export { StateType, PropsType };
