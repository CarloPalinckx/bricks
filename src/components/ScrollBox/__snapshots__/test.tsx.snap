// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ScrollBox should show a bottom shadow when scrolling up 1`] = `
.c0 {
  background: linear-gradient(to bottom,rgba(0,0,0,0) 20%,rgba(0,0,0,0.1) 100%);
  height: 6px;
  position: absolute;
  bottom: 0;
  left: 0;
  right: 0;
}

<styled.div>
  <div
    className="c0"
  />
</styled.div>
`;

exports[`ScrollBox should show a top shadow when scrolling down 1`] = `
.c0 {
  background: linear-gradient(to bottom,rgba(0,0,0,0.1) 20%,rgba(0,0,0,0) 100%);
  height: 6px;
  position: absolute;
  top: 0;
  left: 0;
  right: 0;
}

<styled.div>
  <div
    className="c0"
  />
</styled.div>
`;
